dnf5-plugins/repoclosure_plugin/repoclosure.cpp

```diff
@@ -0,0 +1,216 @@
+    context.set_load_system_repo(false);
+    // filelists needed because there are packages in repos with file requirements
+    context.base.get_config().get_optional_metadata_types_option().add_item(libdnf::METADATA_TYPE_FILELISTS);
+    context.set_load_available_repos(Context::LoadAvailableRepos::ENABLED);
+}
+
+void RepoclosureCommand::run() {
+    auto & ctx = get_context();
+
+    // available_query is the set of packages used to satisfy dependencies
+    // to_check_query is the set of packages we are checking the dependencies of
+    // in case that `--newest` was used, start with an empty query
+    bool newest_used = newest->get_value();
+    libdnf::rpm::PackageQuery available_query(ctx.base, libdnf::sack::ExcludeFlags::APPLY_EXCLUDES, newest_used);
+    libdnf::rpm::PackageQuery to_check_query(ctx.base, libdnf::sack::ExcludeFlags::APPLY_EXCLUDES, newest_used);
+    if (newest_used) {
+        libdnf::repo::RepoQuery repos(ctx.base);
+        repos.filter_enabled(true);
+        for (const auto & repo : repos) {
+            libdnf::rpm::PackageQuery repo_pkgs(ctx.base);
+            repo_pkgs.filter_repo_id({repo->get_id()});
+            repo_pkgs.filter_latest_evr();
+            available_query |= repo_pkgs;
+            to_check_query |= repo_pkgs;
+        }
+    }
+
+    if (!check_repos.empty()) {
+        to_check_query.filter_repo_id(check_repos);
+    }
+
+    if (!arches.empty()) {
+        to_check_query.filter_arch(arches);
+    }
+
+    if (ctx.base.get_config().get_best_option().get_value()) {
+        available_query.filter_latest_evr();
+    }
+
+    if (!pkg_specs.empty()) {
+        libdnf::rpm::PackageQuery to_check_pkgs(ctx.base, libdnf::sack::ExcludeFlags::APPLY_EXCLUDES, true);
+        libdnf::ResolveSpecSettings settings{.with_nevra = true, .with_provides = false, .with_filenames = false};
+        bool specs_resolved = true;
+        for (const auto & spec : pkg_specs) {
+            libdnf::rpm::PackageQuery package_query(to_check_query);
+            auto nevra_pair = package_query.resolve_pkg_spec(spec, settings, true);
+            if (!nevra_pair.first) {
+                specs_resolved = false;
+                std::cerr << libdnf::utils::sformat(_("No match for argument \"{}\"."), spec) << std::endl;
+                continue;
+            }
+            to_check_pkgs |= package_query;
+        }
+        if (!specs_resolved) {
+            throw libdnf::cli::CommandExitError(1, M_("Failed to resolve package specifications."));
+        }
+        to_check_query = to_check_pkgs;
+    }
+
+
+    // cache query results: reldepid -> whether the reldep is satisfied
+    std::unordered_map<int, bool> resolved;
+    std::vector<std::pair<libdnf::rpm::Package, std::vector<std::string>>> unresolved_packages;
+    for (const auto & pkg : to_check_query) {
+        std::vector<std::string> unsatisfied;
+        for (const auto & reldep : pkg.get_requires()) {
+            int reldep_id = reldep.get_id().id;
```

Comment:
Do we want to handle there also file requires? I guess the situation for them is more complicated. And the current behavior includes them compared to the `dnf`. When file patterns are dropped like this, the results seem to be the same:
```
if (libdnf::utils::is_file_pattern(reldep.to_string())) {
    continue;
}
```

Reply:
This PR is 1:1 copy of dnf4 functionality. There are (at least) two fields where it can be enhanced:

- file conflicts (see [Bug 1702652](https://bugzilla.redhat.com/show_bug.cgi?id=1702652) - RFE: extend "dnf repoclosure" with check for file conflicts)
- modularity (see https://github.com/rpm-software-management/dnf5/issues/511)

Reply:
Oh, sorry - I've mixed file conflicts with file provides...

Reply:
I think we should check the file requires (and I was convinced that dnf4 does the same - I need to re-check).

Reply:
I've just tested it running the `dnf repoclosure` and `dnf5 repoclosure`. When omitting the file requires, the output is same on my machine.

Reply:
Hm, this is interesting, because that's why I added explicit `add_item(libdnf::METADATA_TYPE_FILELISTS)` and was testing that the results are the same for dnf4 and dnf5 (except for sorting of unsatisfied package dependencies which dnf4 does not do).
May be there is an issue in libdnf5 regarding loading different sets of repository metadata. Can you remove the repository cache and try again?

Reply:
Yep, you're right. When I remove the cache completely, the repoclosure is working fine with the fresh metadata. Could it be something like that some repositories were marked as expired and refreshed and some were not and therefore some updated requirements are not being provided?

Reply:
Definitely worth a github issue... I'll try to find a reliable reproducer (as the first step to the solution :)
It looks like in some situations the filelists metadata is not loaded (even though requested).